tests/light/functional_tests/filterx/test_filterx.py

```diff
@@ -1534,3 +1534,130 @@ def test_unset_empties(config, syslog_ng):
+    assert "processed" not in file_false.get_stats()
+    assert file_true.read_log() == "bar\n"
+
+
+def test_null_coalesce_use_default_on_error_and_supress_error(config, syslog_ng):
+    (file_true, file_false) = create_config(
+        config, r"""
+        y = "bar";
+        $MSG = len(3.14) ?? y; # in-line expression evaluation is mandatory. the error would be propagated other way
+    """,
+    )
+    syslog_ng.start(config)
+
+    assert file_true.get_stats()["processed"] == 1
+    assert "processed" not in file_false.get_stats()
+    assert file_true.read_log() == "bar\n"
+
+
+def test_null_coalesce_get_subscript_happy_path(config, syslog_ng):
+    (file_true, file_false) = create_config(
+        config, r"""
+        data = json({"foo":"1", "bar":"2", "baz":"3"});
+        def = "default";
+        key = "bar";
+        $MSG = data[key] ?? def;
+    """,
+    )
+    syslog_ng.start(config)
+
+    assert file_true.get_stats()["processed"] == 1
+    assert "processed" not in file_false.get_stats()
+    assert file_true.read_log() == "2\n"
+
+
+def test_null_coalesce_get_subscript_error(config, syslog_ng):
+    (file_true, file_false) = create_config(
+        config, r"""
+        data = json({"foo":"1", "bar":"2", "baz":"3"});
+        def = "default";
+        key = "missing_key";
+        $MSG = data[key] ?? def;
+    """,
+    )
+    syslog_ng.start(config)
+
+    assert file_true.get_stats()["processed"] == 1
+    assert "processed" not in file_false.get_stats()
+    assert file_true.read_log() == "default\n"
+
+
+def test_null_coalesce_use_nested_coalesce(config, syslog_ng):
+    (file_true, file_false) = create_config(
+        config, r"""
+        data = json({"foo":"1", "bar":"2", "baz":"3"});
+        def = "default";
+        key1 = "missing_key1";
+        key2 = "missing_key2";
+        $MSG = data[key1] ?? data[key2] ?? def;
+    """,
+    )
+    syslog_ng.start(config)
+
+    assert file_true.get_stats()["processed"] == 1
+    assert "processed" not in file_false.get_stats()
+    assert file_true.read_log() == "default\n"
+
+
+def test_null_coalesce_use_nested_coalesce_return_mid_match(config, syslog_ng):
+    (file_true, file_false) = create_config(
+        config, r"""
+        data = json({"foo":"1", "bar":"2", "baz":"3"});
+        def = "default";
+        key1 = "missing_key1";
+        key2 = "baz";
+        $MSG = data[key1] ?? data[key2] ?? def;
+    """,
+    )
+    syslog_ng.start(config)
+
+    assert file_true.get_stats()["processed"] == 1
+    assert "processed" not in file_false.get_stats()
+    assert file_true.read_log() == "3\n"
+
+
+def test_null_coalesce_do_not_supress_last_error(config, syslog_ng):
+    (file_true, file_false) = create_config(
+        config, r"""
+        data = json({"foo":"1", "bar":"2", "baz":"3"});
+        def = "default";
+        key1 = "missing_key1";
+        key2 = "missing_key2";
+        key3 = "missing_key3";
+        $MSG = data[key1] ?? data[key2] ?? data[key3];
```

Comment:
It would be nice to see a mix of `??` and `? :` in a TC. It is a corner case which can be handled in different ways, and it should match with the rest of the world, if there is a common way in other languages.